### PR TITLE
51_尹礼春_第三课作业

### PR DESCRIPTION
**Pull request by ylic2018**
@taxic

### CODE REVIEW

File: Lesson-3/assignment/第三题.MD [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+L(Z) = Z + merge(L(K1) + L(K2) + K1K2)<br>
+L(Z) = Z + merge([K1,A,B,O] + [K2,A,C,O] + K1K2)<br>
+L(Z) = Z + K1 + merge([A,B,O] + [K2,A,C,O] + K2)<br>
+L(Z) = Z + K1 + K2 +　merge([A,B,O] + [A,C,O])<br>
+L(Z) = Z + K1 + K2　+ A + merge([B,O] + [C,O])<br>
+L(Z) = Z + K1 + K2 + A　+ B + merge([O] + [C,O])<br>
+L(Z) = Z + K1 + K2 + A　+ B + C + merge([O] + [O])<br>
+L(Z) = Z + K1 + K2 + A　+ B + C + O<br>
+L(Z) = [Z,K1,K2,A,B,C,O]<br>
+
```
Comment (zchcai) [inline Lesson-3/assignment/第三题.MD]: L(O) = [O]
L(A) = [A, O]
L(B) = [B, O]
L(C) = [C, O]
L(K2) = [K2] + merge(L(C), L(A), [C, A])
     = [K2] + merge([C, O], [A, O], [C, A])
     = [K2, C] + merge([O], [A, O], [A])
     = [K2, C, A, O]
L(K1) = [K1] + merge(L(B), L(A), [B, A])
     = [K1] + merge([B, O], [A, O], [B, A])
     = [K1, B] + merge([O], [A, O], [A])
     = [K1, B, A, O]
     
L(Z) = [Z] + merge(L(K2), L(K1), [K2, K1])
    = [Z] + merge([K2, C, A, O], [K1, B, A, O], [K2, K1])
    = [Z, K2] + merge([C, A, O], [K1, B, A, O], [K1])
    = [Z, K2, C] + merge([A, O], [K1, B, A, O], [K1])
    = [Z, K2, C, K1] + merge([A, O], [B, A, O])
    = [Z, K2, C, K1, B] + merge([A, O], [A, O])
    = [Z, K2, C, K1, B, A, O]

File: Lesson-3/assignment/lesson3.sol [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+    //    require(msg.sender == owner);
+    //    _;
+    //}
+    
+    //检查employeeId是否存在
+    modifier employeeExist(address employeeId){
+        Employee storage employee = employees[employeeId];
+        assert(employee.id != 0x0);
+        _;
+    }
```
Comment (zchcai) [inline Lesson-3/assignment/lesson3.sol:39]: try to add employeeNonExist

**Comment by ylic2018** [comment]
@taxic

File: Lesson-3/assignment/lesson3.sol [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+        _partialPaid(employee);
+        uint sa = salary.mul(1 ether);
+        salarysum = salarysum.add(sa).sub(employee.salary);
+        employees[employeeId].salary = sa;
+        employees[employeeId].lastPayday = now;
+    }
+    
+    //由员工自己更换员工地址，需要满足一个付款周期
+    function changePaymentAddress(address employeeId) public  employeeExist(msg.sender) employeeNonExist(employeeId) moreThanpayDuration(employeeId){
+        Employee storage employee = employees[msg.sender];
```
Comment (zchcai) [inline Lesson-3/assignment/lesson3.sol:97]: require new address not zero

Comment (ylic2018) [reply Lesson-3/assignment/lesson3.sol]: modifier employeeNonExist(address employeeId){
        assert(employeeId != 0x0);
        Employee storage employee = employees[employeeId];
        assert(employee.id == 0x0);
        _;
    }

这个modifier里面已经加了不等于0的判断。

**Pull request merged**